File: case-02-final-api-spec-discussion/client_validator.py
#!/usr/bin/env python3
"""
Client validator for GET /dashboard/orders.
Fetches from the mock server and validates the response structure
against the agreed API spec (final_api_spec.yaml).

Usage: python3 client_validator.py [--url URL]
Default URL: http://localhost:8080/dashboard/orders?page=1&page_size=20
"""

import json
import sys
import urllib.request
import urllib.error
from datetime import datetime

MOCK_SERVER_URL = "http://localhost:8080/dashboard/orders?page=1&page_size=20"

VALID_STATUSES = {"pending", "confirmed", "shipped", "delivered", "cancelled", "returned"}

REQUIRED_ORDER_FIELDS = {
    "order_id": str,
    "status": str,
    "total_cents": int,
    "currency": str,
    "created_at": str,
    "total_items_count": int,
    "items": list,
}

REQUIRED_ITEM_FIELDS = {
    "product_id": str,
    "product_name": str,
    "product_image_url": str,
    "unit_price_cents": int,
    "quantity": int,
}

REQUIRED_PAGINATION_FIELDS = {
    "page": int,
    "page_size": int,
    "total_orders": int,
    "total_pages": int,
}


def fail(msg):
    print(f"FAIL: {msg}")
    return False


def validate_iso8601(value, context):
    try:
        datetime.fromisoformat(value.replace("Z", "+00:00"))
        return True
    except (ValueError, AttributeError):
        return fail(f"{context}: '{value}' is not a valid ISO 8601 datetime")


def validate_uri(value, context):
    if not (value.startswith("http://") or value.startswith("https://")):
        return fail(f"{context}: '{value}' is not a valid URI (must start with http:// or https://)")
    return True


def validate_fields(obj, required_fields, context):
    ok = True
    for field, expected_type in required_fields.items():
        if field not in obj:
            fail(f"{context}: missing required field '{field}'")
            ok = False
            continue
        if not isinstance(obj[field], expected_type):
            fail(f"{context}.{field}: expected {expected_type.__name__}, got {type(obj[field]).__name__}")
            ok = False
    return ok


def validate_item(item, context):
    ok = validate_fields(item, REQUIRED_ITEM_FIELDS, context)
    if not ok:
        return False

    if not validate_uri(item["product_image_url"], f"{context}.product_image_url"):
        ok = False

    if item["quantity"] < 1:
        fail(f"{context}.quantity: must be >= 1, got {item['quantity']}")
        ok = False

    return ok


def validate_order(order, index):
    context = f"orders[{index}]"
    ok = validate_fields(order, REQUIRED_ORDER_FIELDS, context)
    if not ok:
        return False

    # Validate status enum
    if order["status"] not in VALID_STATUSES:
        fail(f"{context}.status: '{order['status']}' not in valid statuses {VALID_STATUSES}")
        ok = False

    # Validate created_at is ISO 8601
    if not validate_iso8601(order["created_at"], f"{context}.created_at"):
        ok = False

    # Validate total_items_count >= 1
    if order["total_items_count"] < 1:
        fail(f"{context}.total_items_count: must be >= 1, got {order['total_items_count']}")
        ok = False

    # Validate items array max 3
    if len(order["items"]) > 3:
        fail(f"{context}.items: max 3 items allowed, got {len(order['items'])}")
        ok = False

    # Validate items count consistency
    if order["total_items_count"] < len(order["items"]):
        fail(f"{context}: total_items_count ({order['total_items_count']}) < items length ({len(order['items'])})")
        ok = False

    # Validate each item
    for i, item in enumerate(order["items"]):
        if not validate_item(item, f"{context}.items[{i}]"):
            ok = False

    return ok


def validate_pagination(pagination):
    context = "pagination"
    ok = validate_fields(pagination, REQUIRED_PAGINATION_FIELDS, context)
    if not ok:
        return False

    if pagination["page"] < 1:
        fail(f"{context}.page: must be >= 1, got {pagination['page']}")
        ok = False

    if pagination["page_size"] < 1 or pagination["page_size"] > 50:
        fail(f"{context}.page_size: must be 1-50, got {pagination['page_size']}")
        ok = False

    if pagination["total_orders"] < 0:
        fail(f"{context}.total_orders: must be >= 0, got {pagination['total_orders']}")
        ok = False

    if pagination["total_pages"] < 0:
        fail(f"{context}.total_pages: must be >= 0, got {pagination['total_pages']}")
        ok = False

    return ok


def validate_response(data):
    ok = True

    # Top-level required fields
    if "orders" not in data:
        return fail("Response missing required field 'orders'")
    if "pagination" not in data:
        return fail("Response missing required field 'pagination'")

    if not isinstance(data["orders"], list):
        return fail(f"'orders' must be an array, got {type(data['orders']).__name__}")
    if not isinstance(data["pagination"], dict):
        return fail(f"'pagination' must be an object, got {type(data['pagination']).__name__}")

    # Validate each order
    for i, order in enumerate(data["orders"]):
        if not isinstance(order, dict):
            fail(f"orders[{i}]: expected object, got {type(order).__name__}")
            ok = False
            continue
        if not validate_order(order, i):
            ok = False

    # Validate pagination
    if not validate_pagination(data["pagination"]):
        ok = False

    return ok


def main():
    url = sys.argv[1] if len(sys.argv) > 1 and not sys.argv[1].startswith("--") else None
    if url is None:
        for i, arg in enumerate(sys.argv[1:], 1):
            if arg == "--url" and i < len(sys.argv) - 1:
                url = sys.argv[i + 1]
                break
    if url is None:
        url = MOCK_SERVER_URL

    print(f"Fetching: {url}")

    try:
        req = urllib.request.Request(url, headers={"Accept": "application/json"})
        with urllib.request.urlopen(req, timeout=10) as resp:
            if resp.status != 200:
                print(f"FAIL: Expected HTTP 200, got {resp.status}")
                sys.exit(1)

            content_type = resp.headers.get("Content-Type", "")
            if "application/json" not in content_type:
                print(f"FAIL: Expected Content-Type application/json, got '{content_type}'")
                sys.exit(1)

            body = resp.read().decode("utf-8")

    except urllib.error.URLError as e:
        print(f"FAIL: Could not connect to {url}: {e}")
        sys.exit(1)

    try:
        data = json.loads(body)
    except json.JSONDecodeError as e:
        print(f"FAIL: Response is not valid JSON: {e}")
        sys.exit(1)

    print(f"Response received: {len(data.get('orders', []))} orders")
    print("---")

    if validate_response(data):
        print("---")
        print("PASS: All validations passed.")
        print(f"  Orders: {len(data['orders'])}")
        print(f"  Page: {data['pagination']['page']}/{data['pagination']['total_pages']}")
        sys.exit(0)
    else:
        print("---")
        print("FAIL: Validation errors found (see above).")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: case-02-final-api-spec-discussion/client_validator.js
/**
 * Client Validator for GET /dashboard/orders
 * Project Legacy Phoenix — Frontend validation
 *
 * Fetches from the mock server and validates the response matches
 * the agreed API spec (final_api_spec.yaml). Ensures all fields
 * needed for order card rendering are present and correctly typed.
 *
 * Usage: node client_validator.js [base_url]
 *   base_url defaults to http://localhost:8080
 */

const BASE_URL = process.argv[2] || "http://localhost:8080";
const ENDPOINT = "/dashboard/orders";

const VALID_STATUSES = ["pending", "processing", "shipped", "delivered", "cancelled"];
const VALID_SORTS = ["created_at:desc", "created_at:asc", "total_amount:desc", "total_amount:asc"];

let passCount = 0;
let failCount = 0;

function check(label, condition, detail) {
  if (condition) {
    passCount++;
    console.log(`  PASS: ${label}`);
  } else {
    failCount++;
    console.log(`  FAIL: ${label}${detail ? ` — ${detail}` : ""}`);
  }
}

function isNonEmptyString(val) {
  return typeof val === "string" && val.length > 0;
}

function isPositiveNumber(val) {
  return typeof val === "number" && val > 0;
}

function isNonNegativeInteger(val) {
  return Number.isInteger(val) && val >= 0;
}

function isPositiveInteger(val) {
  return Number.isInteger(val) && val > 0;
}

function isISO8601(val) {
  if (typeof val !== "string") return false;
  const d = new Date(val);
  return !isNaN(d.getTime());
}

function isURL(val) {
  if (typeof val !== "string") return false;
  try {
    new URL(val);
    return true;
  } catch {
    return false;
  }
}

function validateOrderItem(item, index, orderIndex) {
  const prefix = `orders[${orderIndex}].items[${index}]`;
  check(`${prefix}.product_id is a non-empty string`, isNonEmptyString(item.product_id), `got ${JSON.stringify(item.product_id)}`);
  check(`${prefix}.product_name is a non-empty string`, isNonEmptyString(item.product_name), `got ${JSON.stringify(item.product_name)}`);
  check(`${prefix}.thumbnail_url is a valid URL`, isURL(item.thumbnail_url), `got ${JSON.stringify(item.thumbnail_url)}`);
  check(`${prefix}.quantity is a positive integer`, isPositiveInteger(item.quantity), `got ${JSON.stringify(item.quantity)}`);
  check(`${prefix}.unit_price is a positive number`, isPositiveNumber(item.unit_price), `got ${JSON.stringify(item.unit_price)}`);
}

function validateOrder(order, index) {
  const prefix = `orders[${index}]`;

  // Required fields existence
  check(`${prefix}.order_id is a non-empty string`, isNonEmptyString(order.order_id), `got ${JSON.stringify(order.order_id)}`);
  check(`${prefix}.status is a valid enum`, VALID_STATUSES.includes(order.status), `got ${JSON.stringify(order.status)}`);
  check(`${prefix}.created_at is ISO 8601`, isISO8601(order.created_at), `got ${JSON.stringify(order.created_at)}`);
  check(`${prefix}.total_amount is a positive number`, isPositiveNumber(order.total_amount), `got ${JSON.stringify(order.total_amount)}`);
  check(`${prefix}.currency is a non-empty string`, isNonEmptyString(order.currency), `got ${JSON.stringify(order.currency)}`);
  check(`${prefix}.item_count is a non-negative integer`, isNonNegativeInteger(order.item_count), `got ${JSON.stringify(order.item_count)}`);

  // Items array
  check(`${prefix}.items is an array`, Array.isArray(order.items), `got ${typeof order.items}`);
  if (Array.isArray(order.items)) {
    check(`${prefix}.items is non-empty`, order.items.length > 0, "items array is empty");
    check(`${prefix}.item_count matches items.length`, order.item_count === order.items.length, `item_count=${order.item_count} but items.length=${order.items.length}`);

    // Validate items sorted by unit_price DESC
    if (order.items.length > 1) {
      const sorted = order.items.every((item, i) => {
        if (i === 0) return true;
        return item.unit_price <= order.items[i - 1].unit_price;
      });
      check(`${prefix}.items sorted by unit_price DESC`, sorted, "items are not sorted by unit_price descending");
    }

    // Validate each item
    order.items.forEach((item, i) => validateOrderItem(item, i, index));
  }

  // Card rendering check: first item must have thumbnail + name for card display
  if (Array.isArray(order.items) && order.items.length > 0) {
    const firstItem = order.items[0];
    check(`${prefix} card render: first item has thumbnail_url`, isURL(firstItem.thumbnail_url), "missing thumbnail for card");
    check(`${prefix} card render: first item has product_name`, isNonEmptyString(firstItem.product_name), "missing product name for card");
  }
}

function validatePagination(pagination) {
  check("pagination is an object", typeof pagination === "object" && pagination !== null, `got ${typeof pagination}`);
  if (typeof pagination !== "object" || pagination === null) return;

  check("pagination.page is a positive integer", isPositiveInteger(pagination.page), `got ${JSON.stringify(pagination.page)}`);
  check("pagination.page_size is a positive integer", isPositiveInteger(pagination.page_size), `got ${JSON.stringify(pagination.page_size)}`);
  check("pagination.page_size <= 50", typeof pagination.page_size === "number" && pagination.page_size <= 50, `got ${pagination.page_size}`);
  check("pagination.total_count is a non-negative integer", isNonNegativeInteger(pagination.total_count), `got ${JSON.stringify(pagination.total_count)}`);
  check("pagination.total_pages is a non-negative integer", isNonNegativeInteger(pagination.total_pages), `got ${JSON.stringify(pagination.total_pages)}`);
}

async function fetchOrders(params = {}) {
  const url = new URL(`${BASE_URL}${ENDPOINT}`);
  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, value);
  }
  const response = await fetch(url.toString());
  return { status: response.status, data: await response.json() };
}

async function runValidation() {
  console.log(`\nClient Validator — GET ${ENDPOINT}`);
  console.log(`Target: ${BASE_URL}`);
  console.log("=".repeat(60));

  // Test 1: Default request
  console.log("\n[Test 1] Default request (no params)");
  const { status, data } = await fetchOrders();
  check("HTTP status is 200", status === 200, `got ${status}`);
  check("Response has 'orders' array", Array.isArray(data.orders), `got ${typeof data.orders}`);
  check("Response has 'pagination' object", typeof data.pagination === "object", `got ${typeof data.pagination}`);

  if (Array.isArray(data.orders)) {
    check("orders array is non-empty", data.orders.length > 0, "no orders returned");
    check("orders count <= default page_size (20)", data.orders.length <= 20, `got ${data.orders.length}`);

    console.log(`\n  Validating ${data.orders.length} orders...`);
    data.orders.forEach((order, i) => validateOrder(order, i));
  }

  if (data.pagination) {
    console.log("\n  Validating pagination...");
    validatePagination(data.pagination);
  }

  // Test 2: With query params
  console.log("\n[Test 2] Request with params (page=1, page_size=5, sort=total_amount:desc)");
  const { status: s2, data: d2 } = await fetchOrders({ page: 1, page_size: 5, sort: "total_amount:desc" });
  check("HTTP status is 200", s2 === 200, `got ${s2}`);
  if (Array.isArray(d2.orders)) {
    check("orders count <= requested page_size (5)", d2.orders.length <= 5, `got ${d2.orders.length}`);
  }
  if (d2.pagination) {
    check("pagination.page_size reflects request", d2.pagination.page_size === 5, `got ${d2.pagination.page_size}`);
  }

  // Test 3: Status filter
  console.log("\n[Test 3] Request with status filter (status=shipped)");
  const { status: s3, data: d3 } = await fetchOrders({ status: "shipped" });
  check("HTTP status is 200", s3 === 200, `got ${s3}`);
  if (Array.isArray(d3.orders)) {
    const allShipped = d3.orders.every((o) => o.status === "shipped");
    check("All returned orders have status 'shipped'", allShipped, "found orders with non-shipped status");
  }

  // Summary
  console.log("\n" + "=".repeat(60));
  console.log(`RESULTS: ${passCount} passed, ${failCount} failed`);
  if (failCount === 0) {
    console.log("All validations PASSED — response matches spec for card rendering.");
  } else {
    console.log("Some validations FAILED — response does not fully match spec.");
  }
  console.log();

  process.exit(failCount > 0 ? 1 : 0);
}

runValidation().catch((err) => {
  console.error(`\nFATAL: Could not connect to ${BASE_URL}${ENDPOINT}`);
  console.error(`Error: ${err.message}`);
  console.error("Is the mock server running?");
  process.exit(2);
});
